File: BOJ_2667.cpp
//https://www.acmicpc.net/problem/2667
//FloodFill

#include <iostream>
#include <cstdlib>
using namespace std;
int N;
#define MAX ((int)625)
char map[25+10][25+10];
int visited[25+10][25+10];

int direct[4][2]={
	1,0,
	-1,0,
	0,1,
	0,-1
};
int build[MAX+10];
int cnt;
void FloodFill(int y, int x){
	if(y<0||y>=N||x<0||x>=N)return;
	if(visited[y][x]==1)return;
	if(map[y][x]!='1')return;
	visited[y][x]=1;
	map[y][x]='0';
	cnt++;
	for(int i=0;i<4;i++){
		int dy=y+direct[i][0];
		int dx=x+direct[i][1];
		FloodFill(dy,dx);
	}
}

int compare(const void* a, const void* b){
	return *(int *)a-*(int *)b;
}

int main(){

	int count=0;
	int sol=0;
	cin>>N;
	for(int i=0;i<N;i++){
		cin>>map[i];
	}
	for(int y=0;y<N;y++){
		for(int x=0; x<N;x++){
			if(map[y][x]=='1'){
				count++;
				cnt=0;
				FloodFill(y,x);
				build[sol++]=cnt;
			}
		}
	}
	cout<<count<<endl;
	qsort(build,sol,sizeof(build[0]),compare);
	for(int i=0;i<sol;i++){
		cout<<build[i]<<endl;
	}
	return 0;
}

File: BOJ_2667.py
#https://www.acmicpc.net/problem/2667
import sys
N=int(input())
matrix=[[0 for col in range(N)] for low in range(N)]
visited=[[0 for col in range(N)] for low in range(N)]

for i in range(N):
    line=sys.stdin.readline().strip()
    for j,v in enumerate(line):
        matrix[i][j]=int(v)

#print(matrix) # for debugging

direct = [[-1,0],[1,0],[0,-1],[0,1]]

def floodfill(y,x):
    if y>=N or y<0 or x>=N or x<0:
        return
    if visited[y][x] == 1:
        return
    if matrix[y][x] == 0:
        return

    visited[y][x]=1
    global nums
    #global cnt # for debugging
    nums+=1
    #matrix[y][x]=cnt # for debugging

    for i in range(4):
        dy=y+direct[i][0]
        dx=x+direct[i][1]
        floodfill(dy,dx)

cnt=0 # Number of Apartment complex
numlist=[] # Storage nums for each complex
nums=0 # Number of Apartment

for y in range(N):
    for x in range(N):
        if matrix[y][x] == 1 and visited[y][x] == 0:
            cnt+=1
            floodfill(y,x)
            #print(y,x,nums) # for debugging
            numlist.append(nums)
            nums=0
#print(matrix) # for debugging
print(len(numlist))
for n in sorted(numlist):
    print(n)